synapse/api/auth.py

```diff
@@ -224,12 +224,22 @@ async def get_user_by_req(
             shadow_banned = user_info.shadow_banned
 
             # Deny the request if the user account has expired.
-            if self._account_validity_enabled and not allow_expired:
-                if await self.store.is_account_expired(
-                    user_info.user_id, self.clock.time_msec()
+            if not allow_expired:
+                if await self.hs.get_account_validity().user_expired(
```

Comment:
ohh, or better:

we change round the way modules work, so we just have a single `modules` list at the top level of the homeserver config. Each module gets loaded at startup, but after all the handlers. Then the module registers for the hooks it needs (in this case, `account_validity` and additional http resources).

Reply:
> I'm guessing that we're doing this `hs.get_account_validity()` here rather than in the constructor to avoid a dependency loop?

Yes, iirc it's because `hs.get_auth()` is called in the module API's `__init__` function (or because a handler that's listed [here](https://github.com/matrix-org/synapse/blob/2dd3a354b1eb827c2ea54c075bfaa6aa942a1d45/synapse/server.py#L215-L223) does it).

> we change round the way modules work, so we just have a single `modules` list at the top level of the homeserver config. Each module gets loaded at startup, but after all the handlers. Then the module registers for the hooks it needs (in this case, `account_validity` and additional http resources).

I was thinking of that earlier, and I really like the idea. It would also have the added bonus of making documenting module development easier since you wouldn't need a different documentation for every different type of module.

Reply:
>  I was thinking of that earlier, and I really like the idea.

Yeah. If you fancy making that change here, I think it will be very worthwhile (and hopefully not too much work).

Reply:
I think I'm happy to do this change but given it changes some things about other modules as well it might be better to do it in a separate PR.

Reply:
I think splitting up this PR is a good idea - it's got a bit unweildly.

I wouldn't leap into changing the way other modules work right now. We'll need to keep the existing interfaces around for quite a while, even if they are deprecated.

Reply:
Resolving the thread as I believe the conclusion is that this will be done in a future PR (and I have the conn on making that happen)

Reply:
I'm reopening this. Whilst we should leave the *existing* module hooks in place for a future PR, I don't think we should add any *new* module hooks that assume single-function modules, so I'm afraid I'd like to see changes before this PR lands.